Assets/Editor/SchedulerEditorTest.cs

```diff
@@ -0,0 +1,311 @@
+        {
+            Assert.That(tally, Is.EqualTo(expectedTally[t]));
+            scheduler.Update(1.0f);
+        }
+    }
+
+    [Test]
+    public void SchedulerTimeToNextEventTest()
+    {
+        Assert.That(scheduler.Events.Count, Is.EqualTo(0));
+
+        ScheduledEvent evt1 = new ScheduledEvent(
+            "test",
+            callback,
+            2.0f,
+            true,
+            1);
+        ScheduledEvent evt2 = new ScheduledEvent(
+            "test",
+            callback,
+            3.0f,
+            true,
+            1);
+
+        scheduler.RegisterEvent(evt2);
+        scheduler.Update(0); // updates the event list
+        Assert.That(scheduler.TimeToNextEvent(), Is.EqualTo(3));
+        scheduler.RegisterEvent(evt1);
+        scheduler.Update(0); // updates the event list
+        Assert.That(scheduler.TimeToNextEvent(), Is.EqualTo(2));
+    }
+
+    [Test]
+    public void SchedulerClearFinishedEventsTest()
+    {
+        Assert.That(scheduler.Events.Count, Is.EqualTo(0));
+
+        ScheduledEvent evt1 = new ScheduledEvent(
+            "test",
+            callback,
+            2.0f,
+            true,
+            1);
+        ScheduledEvent evt2 = new ScheduledEvent(
+            "test",
+            callback,
+            3.0f,
+            true,
+            1);
+
+        scheduler.RegisterEvent(evt1);
+        scheduler.RegisterEvent(evt2);
+        scheduler.Update(0); // updates the event list
+
+        evt2.Stop();
+        Assert.That(scheduler.IsRegistered(evt2), Is.True);
+        scheduler.Update(0); // updates the event list
+        Assert.That(scheduler.IsRegistered(evt2), Is.False);
+
+        evt1.Stop();
+        Assert.That(scheduler.IsRegistered(evt1), Is.True);
+        scheduler.Update(0); // updates the event list
+        Assert.That(scheduler.IsRegistered(evt1), Is.False);
+    }
+
+    [Test]
+    public void SchedulerLuaEventTest()
+    {
+        ScheduledEvent evt = new ScheduledEvent(PrototypeManager.SchedulerEvent.GetPrototype("ping_log_lua"), 1.0f, 1.0f, false, 1);
+
+        scheduler.RegisterEvent(evt);
+
+        scheduler.Update(1);
+    }
+
+    [Test]
+    public void SchedulerEventsDoNotModifyEventsListDuringUpdateTest()
+    {
+        Assert.That(scheduler.Events.Count, Is.EqualTo(0));
+
+        int numEventsInList = 0;
+
+        // generic event that expires at time 2s
+        ScheduledEvent evt1 = new ScheduledEvent(
+            "test",
+            callback,
+            2.0f,
+            false,
+            1);
+
+        // event which tries to purge the event list at 5s
+        ScheduledEvent evt2 = new ScheduledEvent(
+            "test",
+            (evt) => scheduler.DeregisterEvent(evt1),
+            5.0f,
+            false,
+            1);
+
+        // event which counts the events in the list at 7s
+        ScheduledEvent evt3 = new ScheduledEvent(
+            "test",
+            (evt) => (numEventsInList = scheduler.Events.Count),
```

Comment:
this must be this ->  `(evt) => numEventsInList = scheduler.Events.Count,`
Without brackets